cluster/gce/delete-stranded-ingress-resources.sh

```diff
@@ -0,0 +1,133 @@
+#!/bin/bash
+
+# Copyright 2015 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# A utility for cleaning up GCE networking resources created through GCE Ingress
+# controller upon deletion of the cluster before deletion of the Ingress.
+set -euo pipefail
+IFS=$'\n\t'
+
+log() {
+	echo "$@" >&2
+}
+
+PROJECT="${PROJECT:?Required variable PROJECT.}"
+
+gcloud() {
+	command gcloud --project="${PROJECT}" "$@"
+}
+
+v_gcloud() {
+	echo "+ gcloud --project=${PROJECT}" "$@"
+	(
+		set +e
+		command gcloud --project="${PROJECT}" "$@"
+	)
+}
+
+has_health_status() {
+	local backend
+	backend="$1"
+
+	local out
+	out="$(gcloud compute backend-services get-health "${backend}" \
+		--global --format='value(status.healthStatus)')"
+	[[ -n "$out" ]]
+}
+
+cleanup_resources() {
+	local backend backend_id
+	backend="$1"
+	backend_id="$(echo "${backend}" | grep -Eo '\-\-([0-9a-f]+)' /dev/stdin | sed 's/^--//')"
+
+	local health_check firewall url_map target_http_proxy target_https_proxy \
+		forwarding_rule instance_groups
+
+	firewall="k8s-fw-l7--${backend_id}"
+
+	url_map="$(gcloud compute url-maps list --format='value(name)' \
+		--filter=defaultService~"${backend}")"
+	if [[ -n "${url_map}" ]]; then
+		forwarding_rule="${url_map/-um-/-fw-}"
+		target_http_proxy="$(gcloud compute target-http-proxies list	\
+			--format='value(name)' --filter=urlMap~"${url_map}")"
+		target_https_proxy="$(gcloud compute target-https-proxies list \
+			--format='value(name)' --filter=urlMap~"${url_map}")"
+	fi
+
+	health_check="$(gcloud compute backend-services describe \
+		--global "${backend}" --format='value(healthChecks)')"
+
+	instance_groups=$(gcloud compute backend-services describe \
+		--global "${backend}" --format='value(backends.group)' | tr \; '\n')
+
+	log "    FIREWALL=${firewall}"
+	log "    HEALTH_CHECK=${health_check}"
+	log "    URL_MAP=${url_map}"
+	log "    TARGET_HTTP_PROXY=${target_http_proxy}"
+	log "    TARGET_HTTPS_PROXY=${target_https_proxy}"
+	log "    FORWARDING_RULE=${forwarding_rule}"
```

Comment:
What about static IP, SSL cert? Later on we may have NEG resource as well.

FYI we implemented a similar function in e2e :)
https://github.com/kubernetes/kubernetes/blob/97798c4f3d02033486b75b83480908bd1cb651d5/test/e2e/framework/ingress_utils.go#L947-L974

Reply:
Static IP: we probably shouldn't delete it because the user might want to hold on to the IP.

SSL cert: +1 We should clean this up.

I'll take a look at the code.